From 54ad71177ade90ba5b66b9d1a642927579ec9b9f Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Fri, 3 Apr 2009 13:05:43 +0200
Subject: [PATCH] OOXML:  Set sheet's IsUndoEnabled before invoking protect.  Fixes n#481317.

ScTableSheetObj::protect () [or actually ScDocFunc::Protect] only
works if UNDO is enabled, otherwise it's a no-op.  Enabling undo
before calling xProtectable->protect () fixes this.
---
 oox/source/xls/worksheetsettings.cxx |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git oox/source/xls/worksheetsettings.cxx oox/source/xls/worksheetsettings.cxx
index d30106a..fc12fe7 100644
--- oox/source/xls/worksheetsettings.cxx
+++ oox/source/xls/worksheetsettings.cxx
@@ -29,8 +29,10 @@
  ************************************************************************/
 
 #include "oox/xls/worksheetsettings.hxx"
+
 #include <com/sun/star/util/XProtectable.hpp>
 #include "oox/helper/attributelist.hxx"
+#include "oox/helper/propertyset.hxx"
 #include "oox/helper/recordinputstream.hxx"
 #include "oox/xls/biffinputstream.hxx"
 #include "oox/xls/pagesettings.hxx"
@@ -45,6 +47,8 @@ using ::com::sun::star::util::XProtectable;
 namespace oox {
 namespace xls {
 
+using namespace ::com::sun::star;
+
 // ============================================================================
 
 namespace {
@@ -314,6 +318,9 @@ void WorksheetSettings::finalizeImport()
     if( maSheetProt.mbSheet ) try
     {
         Reference< XProtectable > xProtectable( getSheet(), UNO_QUERY_THROW );
+        /* ScTableSheetObj::protect () [or actually ScDocFunc::Protect]
+           only works if UNDO is enabled, otherwise it's a no-op.  */
+        PropertySet( getXSpreadsheet() ).setProperty( CREATE_OUSTRING( "IsUndoEnabled" ), uno::makeAny( true ) );
         xProtectable->protect( OUString() );
     }
     catch( Exception& )
